lib/sles4sap/sap_deployment_automation_framework/deployment_connector.pm

```diff
@@ -213,45 +212,31 @@ Deployment ID returned from both jobs: 123456 - because it matches with existing
 
 sub find_deployment_id {
     my (%args) = @_;
-    return get_var('SDAF_DEPLOYMENT_ID') if get_var('SDAF_DEPLOYMENT_ID');
-    return get_var('SDAF_GRANDPARENT_ID') if get_var('SDAF_GRANDPARENT_ID');
-
     $args{deployer_resource_group} //= get_required_var('SDAF_DEPLOYER_RESOURCE_GROUP');
-    my @check_list = (get_current_job_id(), @{get_parent_ids()});
-
-    # For SDAF cleanup job it needs its grandparent's id
-    my @parents = get_parent_ids();
-    my $parent_id = $parents[0][0];
-    if (looks_like_number($parent_id) && @parents == 1) {
-        # Get the SDAF deployment job ID, here it is set via SDAF_GRANDPARENT_ID and saved in vars.json
-        my $vars_json = 'vars.json';
-        my $openqa_url = get_required_var('OPENQA_URL');
-        my $url = Mojo::URL->new("https://$openqa_url/tests/$parent_id/file/$vars_json") or die "No $vars_json found";
-        my $ua = Mojo::UserAgent->new;
-        $ua->insecure(1);
-        my $tx = $ua->get($url);
-        my $json = $tx->res->json;
-        my $id = $json->{SDAF_GRANDPARENT_ID};
-        if ($id) {
-            @check_list = ($id);
-            record_info("SDAF grandparent id of cleanup job:$id");
-        }
-    }
 
-    diag("Job IDs found", Dumper(@check_list));
-    my @ids_found;
-    for my $deployment_id (@check_list) {
-        my $vm_name =
-          get_deployer_vm_name(deployer_resource_group => $args{deployer_resource_group}, deployment_id => $deployment_id);
-        push(@ids_found, $deployment_id) if $vm_name;
-
-        # Set SDAF_GRANDPARENT_ID, SDAF cleanup job needs this id for cleanup
-        set_var('SDAF_GRANDPARENT_ID', $deployment_id) if $vm_name;
+    # Lists VMs in cloud which contain 'deployment_id' tag and displays only tag values => list of all deployments
+    my @cloud_deployments = @{az_vm_list(
+            resource_group => $args{deployer_resource_group}, query => '[?tags.deployment_id].tags.deployment_id')};
+    my @found_deployments;
+    # check current job first
+    my @to_check = get_current_job_id;
+
+    while (my $id = shift(@to_check)) {
+        diag("Checking ID: $id");
+        diag("Deployments in cloud:\n" . join("\n", @cloud_deployments));
+        push @to_check, @{get_parent_ids(job_id => $id)} if get_parent_ids(job_id => $id);
+        diag("IDs to be checked:\n" . join("\n", @to_check));
+
+        push @found_deployments, $id if grep(/^$id$/, @cloud_deployments);
+        diag("Matching deployments:\n" . join("\n", @found_deployments));
```

Comment:
We discussed this already, but perhaps it would be a good idea to add a counter to limit the number of cycles in the loop as a way to avoid potential infinite loops (which I don't expect either because we won't have infinite jobs chained)